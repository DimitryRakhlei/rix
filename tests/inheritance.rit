Rectangle :: BaseType
  w = int
  h = int
  
  ::: int w, int h
    $.w = w
    $.h = h
  
  int : area
    -> w * h

Square :: Rectangle
  ::: int edge
    w = edge
    h = edge


r = Rectangle 5, 10
s = Square 16

print r area
print s area
