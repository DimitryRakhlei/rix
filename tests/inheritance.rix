Rectangle :: BaseType
  w = int
  h = int
  
  $(int w, int h)
    $.w = w
    $.h = h
  
  area->int() = w * h

  perimeter->int() = ( w + h ) *2


NamedRectangle :: Rectangle

  name = String

  $(int w, int h, String name)
    $.w = w
    $.h = h
    $.name = name

  nameArea->String() = "NamedRectangle "+ name + " has an area of "+$.area()


#r = Rectangle (5, 10)
#s = NamedRectangle (3, 4, "Cornelius")

print ("My name is " + s.name)
print ("My area is " + r.area() + " and perimeter is " + r.perimeter())
print (s.name +"'s" + " area is " + s.area())
print (s.nameArea())
