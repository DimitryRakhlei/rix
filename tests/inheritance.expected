My name is Cornelius
My area is 50 and perimeter is 30
Cornelius's area is 12
NamedRectangle Cornelius has an area of 12
